### ISSUE TITLE
KafkaSparkStreamingSpec testsuite does not pass.

### ISSUE THREAD

**Issue by huylv** [miguno/kafka-storm-starter#4]
Hi,
I have problem running `./sbt test`:

```
[info]   Scenario: User creates a Spark Streaming job that reads from and writes to Kafka *** FAILED ***
[info]   0 was not equal to 3 (KafkaSparkStreamingSpec.scala:237)
[info]     Given a ZooKeeper instance
[info]     And a Kafka broker instance
[info]     And some tweets
[info]     And a synchronous Kafka producer app that writes to the topic KafkaTopic(testing-input,1,1,{})
[info]     And a single-threaded Kafka consumer app that reads from topic KafkaTopic(testing-output,1,1,{}) and Avro-decodes the incoming data
[info]     When I Avro-encode the tweets and use the Kafka producer app to sent them to Kafka
[info]     And I run a streaming job that reads tweets from topic KafkaTopic(testing-input,1,1,{}) and writes them as-is to topic KafkaTopic(testing-output,1,1,{})
[info]     Then the Spark Streaming job should consume all tweets from Kafka
2014-10-09 19:43:09 org.apache.spark.streaming.scheduler.ReceiverTracker [ERROR] Deregistered receiver for stream 0: Error starting receiver 0 - org.I0Itec.zkclient.exception.ZkTimeoutException: Unable to connect to zookeeper server within timeout: 1000
    at org.I0Itec.zkclient.ZkClient.connect(ZkClient.java:895)
    at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:98)
    at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:84)
    at kafka.consumer.ZookeeperConsumerConnector.connectZk(ZookeeperConsumerConnector.scala:156)
    at kafka.consumer.ZookeeperConsumerConnector.<init>(ZookeeperConsumerConnector.scala:114)
    at kafka.consumer.ZookeeperConsumerConnector.<init>(ZookeeperConsumerConnector.scala:128)
    at kafka.consumer.Consumer$.create(ConsumerConnector.scala:89)
    at org.apache.spark.streaming.kafka.KafkaReceiver.onStart(KafkaInputDStream.scala:97)
    at org.apache.spark.streaming.receiver.ReceiverSupervisor.startReceiver(ReceiverSupervisor.scala:121)
    at org.apache.spark.streaming.receiver.ReceiverSupervisor.start(ReceiverSupervisor.scala:106)
    at org.apache.spark.streaming.scheduler.ReceiverTracker$ReceiverLauncher$$anonfun$9.apply(ReceiverTracker.scala:264)
    at org.apache.spark.streaming.scheduler.ReceiverTracker$ReceiverLauncher$$anonfun$9.apply(ReceiverTracker.scala:257)
    at org.apache.spark.SparkContext$$anonfun$runJob$4.apply(SparkContext.scala:1121)
    at org.apache.spark.SparkContext$$anonfun$runJob$4.apply(SparkContext.scala:1121)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:62)
    at org.apache.spark.scheduler.Task.run(Task.scala:54)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:177)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)

2014-10-09 19:43:09 org.apache.spark.streaming.receiver.ReceiverSupervisorImpl [ERROR] Stopped executor with error: org.I0Itec.zkclient.exception.ZkTimeoutException: Unable to connect to zookeeper server within timeout: 1000
2014-10-09 19:43:09 org.apache.spark.executor.Executor [ERROR] Exception in task 0.0 in stage 0.0 (TID 0)
org.I0Itec.zkclient.exception.ZkTimeoutException: Unable to connect to zookeeper server within timeout: 1000
    at org.I0Itec.zkclient.ZkClient.connect(ZkClient.java:895) ~[zkclient-0.4.jar:0.4]
    at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:98) ~[zkclient-0.4.jar:0.4]
    at org.I0Itec.zkclient.ZkClient.<init>(ZkClient.java:84) ~[zkclient-0.4.jar:0.4]
    at kafka.consumer.ZookeeperConsumerConnector.connectZk(ZookeeperConsumerConnector.scala:156) ~[kafka_2.10-0.8.1.1.jar:na]
    at kafka.consumer.ZookeeperConsumerConnector.<init>(ZookeeperConsumerConnector.scala:114) ~[kafka_2.10-0.8.1.1.jar:na]
    at kafka.consumer.ZookeeperConsumerConnector.<init>(ZookeeperConsumerConnector.scala:128) ~[kafka_2.10-0.8.1.1.jar:na]
    at kafka.consumer.Consumer$.create(ConsumerConnector.scala:89) ~[kafka_2.10-0.8.1.1.jar:na]
    at org.apache.spark.streaming.kafka.KafkaReceiver.onStart(KafkaInputDStream.scala:97) ~[spark-streaming-kafka_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.streaming.receiver.ReceiverSupervisor.startReceiver(ReceiverSupervisor.scala:121) ~[spark-streaming_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.streaming.receiver.ReceiverSupervisor.start(ReceiverSupervisor.scala:106) ~[spark-streaming_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.streaming.scheduler.ReceiverTracker$ReceiverLauncher$$anonfun$9.apply(ReceiverTracker.scala:264) ~[spark-streaming_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.streaming.scheduler.ReceiverTracker$ReceiverLauncher$$anonfun$9.apply(ReceiverTracker.scala:257) ~[spark-streaming_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.SparkContext$$anonfun$runJob$4.apply(SparkContext.scala:1121) ~[spark-core_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.SparkContext$$anonfun$runJob$4.apply(SparkContext.scala:1121) ~[spark-core_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:62) ~[spark-core_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.scheduler.Task.run(Task.scala:54) ~[spark-core_2.10-1.1.0.jar:1.1.0]
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:177) ~[spark-core_2.10-1.1.0.jar:1.1.0]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_65]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_65]
    at java.lang.Thread.run(Thread.java:745) [na:1.7.0_65]
2014-10-09 19:43:09 org.apache.spark.scheduler.TaskSetManager [ERROR] Task 0 in stage 0.0 failed 1 times; aborting job
```

Running the testsuite within IntelliJ returns another exception:

```
As a user of Spark Streaming


I want to read Avro-encoded data from Kafka


so that I can quickly build Kafka<->Spark Streaming data flows


Exception encountered when invoking run on a nested suite - class "javax.servlet.FilterRegistration"'s signer information does not match signer information of other classes in the same package
java.lang.SecurityException: class "javax.servlet.FilterRegistration"'s signer information does not match signer information of other classes in the same package
    at java.lang.ClassLoader.checkCerts(ClassLoader.java:952)
    at java.lang.ClassLoader.preDefineClass(ClassLoader.java:666)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:794)
    at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
    at java.net.URLClassLoader.defineClass(URLClassLoader.java:449)
    at java.net.URLClassLoader.access$100(URLClassLoader.java:71)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
    at org.eclipse.jetty.servlet.ServletContextHandler.<init>(ServletContextHandler.java:136)
    at org.eclipse.jetty.servlet.ServletContextHandler.<init>(ServletContextHandler.java:129)
    at org.eclipse.jetty.servlet.ServletContextHandler.<init>(ServletContextHandler.java:98)
    at org.apache.spark.ui.JettyUtils$.createServletHandler(JettyUtils.scala:98)
    at org.apache.spark.ui.JettyUtils$.createServletHandler(JettyUtils.scala:89)
    at org.apache.spark.ui.WebUI.attachPage(WebUI.scala:67)
    at org.apache.spark.ui.WebUI$$anonfun$attachTab$1.apply(WebUI.scala:60)
    at org.apache.spark.ui.WebUI$$anonfun$attachTab$1.apply(WebUI.scala:60)
    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
    at org.apache.spark.ui.WebUI.attachTab(WebUI.scala:60)
    at org.apache.spark.ui.SparkUI.initialize(SparkUI.scala:66)
    at org.apache.spark.ui.SparkUI.<init>(SparkUI.scala:60)
    at org.apache.spark.ui.SparkUI.<init>(SparkUI.scala:42)
    at org.apache.spark.SparkContext.<init>(SparkContext.scala:223)
    at org.apache.spark.streaming.StreamingContext$.createNewSparkContext(StreamingContext.scala:555)
    at org.apache.spark.streaming.StreamingContext.<init>(StreamingContext.scala:75)
    at com.miguno.kafkastorm.spark.KafkaSparkStreamingSpec.prepareSparkStreaming(KafkaSparkStreamingSpec.scala:77)
    at com.miguno.kafkastorm.spark.KafkaSparkStreamingSpec.beforeEach(KafkaSparkStreamingSpec.scala:47)
    at org.scalatest.BeforeAndAfterEach$class.beforeEach(BeforeAndAfterEach.scala:154)
    at com.miguno.kafkastorm.spark.KafkaSparkStreamingSpec.beforeEach(KafkaSparkStreamingSpec.scala:30)
    at org.scalatest.BeforeAndAfterEach$class.beforeEach(BeforeAndAfterEach.scala:173)
    at com.miguno.kafkastorm.spark.KafkaSparkStreamingSpec.beforeEach(KafkaSparkStreamingSpec.scala:30)
    at org.scalatest.BeforeAndAfterEach$class.runTest(BeforeAndAfterEach.scala:253)
    at com.miguno.kafkastorm.spark.KafkaSparkStreamingSpec.runTest(KafkaSparkStreamingSpec.scala:30)
    at org.scalatest.FeatureSpecLike$$anonfun$runTests$1.apply(FeatureSpecLike.scala:267)
    at org.scalatest.FeatureSpecLike$$anonfun$runTests$1.apply(FeatureSpecLike.scala:267)
    at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
    at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
    at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:390)
    at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:427)
    at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
    at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
    at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
    at org.scalatest.FeatureSpecLike$class.runTests(FeatureSpecLike.scala:267)
    at org.scalatest.FeatureSpec.runTests(FeatureSpec.scala:1836)
    at org.scalatest.Suite$class.run(Suite.scala:1424)
    at org.scalatest.FeatureSpec.org$scalatest$FeatureSpecLike$$super$run(FeatureSpec.scala:1836)
    at org.scalatest.FeatureSpecLike$$anonfun$run$1.apply(FeatureSpecLike.scala:309)
    at org.scalatest.FeatureSpecLike$$anonfun$run$1.apply(FeatureSpecLike.scala:309)
    at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
    at org.scalatest.FeatureSpecLike$class.run(FeatureSpecLike.scala:309)
    at org.scalatest.FeatureSpec.run(FeatureSpec.scala:1836)
    at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
    at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
    at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
    at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
    at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
    at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
    at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
    at org.scalatest.tools.Runner$.run(Runner.scala:883)
    at org.scalatest.tools.Runner.run(Runner.scala)
    at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:141)
    at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:32)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134)


Process finished with exit code 0
```

Please help.

Thanks,
Huy

**Comment by miguno** [comment]
Hmm, I can't reproduce either of the errors you report, Huy.

Can you provide more details about your environment?  For example, which OS, which Java version, whether you are running in a VM or not, whether you have modified the build, etc.

The first error is a timeout when Spark Streaming is trying to connect to ZooKeeper:

```
2014-10-09 19:43:09 org.apache.spark.streaming.scheduler.ReceiverTracker [ERROR] Deregistered receiver for stream 0: Error starting receiver 0 - org.I0Itec.zkclient.exception.ZkTimeoutException: Unable to connect to zookeeper server within timeout: 1000
```

You can try increasing the timeout in `KafkaSparkStreamingSpec` if you think that you're machine is too slow:

```
        val kafkaParams = Map[String, String](
          "zookeeper.connect" -> kafkaZkCluster.zookeeper.connectString,
          "group.id" -> sparkStreamingConsumerGroup,
          // CAUTION: Spark's use of auto.offset.reset is DIFFERENT from Kafka's behavior!
          // https://issues.apache.org/jira/browse/SPARK-2383
          "auto.offset.reset" -> "smallest", // must be compatible with when/how we are writing the input data to Kafka
          "zookeeper.connection.timeout.ms" -> "1000")
```

Here, change "zookeeper.connection.timeout.ms" from "1000" (1 sec) to "10000" (10 secs).

Alternatively, maybe ZK is not starting on your machine, and hence the timeout?  You may want to increase logging in `src/test/resources/logback-test.xml` and see whether that helps you to pinpoint the problem:

``` xml
  <root level="ERROR">     <!--  <<< change ERROR to DEBUG -->
    <appender-ref ref="A1"/>
  </root>
```

You may also want to read:
- [SPARK-1949](https://issues.apache.org/jira/browse/SPARK-1949): Servlet 2.5 vs 3.0 conflict in SBT build
- http://wpcertification.blogspot.ch/2014/08/wordcount-program-built-using-apache.html (describes the "signer information does not match" problem and a workaround through dependency excludes)

I won't comment on the IDEA issue yet.

**Comment by huylv** [comment]
Hi Michael,

I've just tried to pull again from git and this time it works, both from command line and within IntelliJ. Strange thing is that yesterday I didn't even touch a file before running `./sbt test`, just a `git clone` and then `./sbt test`. I have no idea what's going on. The only difference is that I restarted Chrome before pulling from git (well, there's maybe nothing to do with Chrome, or maybe Chrome ate up all the memory and made Zookeeper go timeout?). I really appreciate your help, however.

By the way, I'm using Oracle Java 1.7.0_65 on Mavericks and everything's running directly from my machine.

Huy
